core/src/main/java/com/linecorp/armeria/common/PublisherBasedHttpResponse.java

```diff
@@ -25,4 +30,77 @@
     PublisherBasedHttpResponse(Publisher<? extends HttpObject> publisher) {
         super(publisher);
     }
+
+    static PublisherBasedHttpResponse from(ResponseHeaders headers,
+                                           Publisher<? extends HttpData> contentPublisher) {
+        return new PublisherBasedHttpResponse(new HeadersAndContentProcessor(headers, contentPublisher));
+    }
+
+    static final class HeadersAndContentProcessor implements Processor<HttpData, HttpObject> {
+
+        private final ResponseHeaders headers;
+        @Nullable
+        private Subscriber<? super HttpObject> subscriber;
+        @Nullable
+        private Subscription contentSubscription;
+
+        HeadersAndContentProcessor(ResponseHeaders headers, Publisher<? extends HttpData> contentPublisher) {
+            this.headers = headers;
+            contentPublisher.subscribe(this);
```

Comment:
This doesn't work:
```java
final HttpResponse res = HttpResponse.of(ResponseHeaders.of(200), Mono.empty());
res.aggregate().join();
```
How about subscribing to `contentPublisher` after `PublisherBasedHttpResponse.subscribe(...)` is called?

Reply:
If we `contentPublisher.subscribe` after `PublisherBasedHttpResponse.subscribe` the subscriber starts requesting before we get `contentSubscription`. I figured the problem was for `Mono.empty`, `onComplete` is called immediately after subscribing, hence `subscriber` is `null`. I handled this case in the latest commit 😄

Reply:
`Subscriber.onError(t)` can also be called before this processor is subscribed. https://github.com/reactive-streams/reactive-streams-jvm#2.10
Also we should use volatile for `contentCompleted` because more than one thread might access the field. 😄

Reply:
> I figured the problem was for Mono.empty, onComplete is called immediately after subscribing, hence subscriber is null

This bug was made by my previous work.
I added a fix to https://github.com/line/armeria/pull/3254/files#diff-d397601b7ec794ee0cac14520792616a29132a6118e337975c86ad5d896b032cR141-R144

Reply:
I think it's not related to the bug. 😄 
`onError()` and `onComplete()` can be called without preceding `subscription.request()`.
https://github.com/reactive-streams/reactive-streams-jvm#2.10
So we shouldn't do `assert subscriber != null` in those methods.